### PR TITLE
Keys API changes

### PR DESCRIPTION
**Pull request by zeripath**
The results of the API for both the repo ssh keys and the user ssh keys lacks several important features.

This pull request adjusts the API to allow reporting of the key id for the keys reported by  `/repos/{owner}/{repo}/keys` and the owner, keytype and read_only type for the keys reported by `/user/keys`.

(Of note `/user/keys` does not only just return user keys but also deploy keys.)

NB: This pull request requires: https://github.com/go-gitea/go-sdk/pull/121

### CODE REVIEW

**Comment by zeripath** [comment]
I've made a few more changes/improvements to this pull-request.

* in `/user/keys` or `/users/:username/keys` User keys get information about their owner if the requester is the owner or an admin. All keys get told which type of key they are.
* in `/repos/:owner/:repo/keys` the fingerprint and the keyID are returned. Information about the real repository the key is attached to is given if the current user is the owner of the repository or an admin.
* `/user/keys?fingerprint=` will search for a key by fingerprint
* `/repos/:owner/:repo/keys?fingerprint=&key_id=` will search the provided repo keys for the key with the given key_id and/or fingerprint.

The main benefit of allowing searching by fingerprint is that it makes it possible to implement an SSH AuthorizedKeysCommand à la #1870. I would have implemented searching by key content but the simplest naïve implementation would allow listing of all keys due to the way Gitea stores the content of keys - it may not be needed in any case.

**Comment by zeripath** [comment]
Heya, any thoughts about this request? It should be possible to implement searching by key content if it was required.

**Comment by lafriks** [comment]
Need to wait for one more maintainer to approve sdk PR

**Comment by zeripath** [comment]
Ugh. It seems that I got go-gitea/go-sdk#121 out of sync with this request so another change to the SDK is needed. I've put another request in for that go-gitea/go-sdk#125 - if that's pulled I can adjust the first changeset and this might finally build!

**Comment by lafriks** [comment]
SDK PR is merged, please update

File: models/ssh_key.go
Comment (lafriks) [inline models/ssh_key.go]: I think this is oracle syntax for escaping percent sign.
I think final qerry condition should be (this should work at least in sqlite, mysql, postgres and mssql):
`col like '%5\%%' escape '\'`
Through I don't know if xorm does support this (@lunny maybe you know?)

**Review by lafriks — CHANGES_REQUESTED** [review]

File: models/ssh_key.go
Comment (zeripath) [reply models/ssh_key.go]: Xorm doesn't support this, but we can use builder.Expr instead.  If you're sure that the ESCAPE keyword is allowed then I'll put that in. 

Is there a better way of doing these tests than using LIKE?

Comment (zeripath) [reply models/ssh_key.go]: OK, 4-5 attempts later I think I have a sensible working escaping function and working LIKE expression...

**Comment by zeripath** [comment]
Do you think the changes to the LIKE are sufficient?

**Comment by HoffmannP** [comment]
I'm afraid we might get into a mess.
My PR #5199 will probably be updating gitea based on my already merged changes to the SDK. But it couldn't be merged until now because of your merged changes to the SDK (affecting keys) that weren't merged into gitea.
So when my PR is approved&merged (after updating, also containing your changes) your PR might get outdated/superfluous even though it could've already been accepted.
The second way would be that your PR ist approved first. Then I might have to change my PR again. But I think that would be the cleaner way.

**Comment by zeripath** [comment]
Hi!

Ugh. What changes are needed to this request to match your request? I don't mind updating, although I have been waiting a while. My changes to the SDK are in already too... 

@lunny @lafriks is this request now ready to go in? Or would you prefer that #5199 go in first and I update to match that?

**Comment by zeripath** [comment]
Hi!

Ugh. What changes are needed to this request to match your request? I don't mind updating, although I have been waiting a while. My changes to the SDK are in already too... 

@lunny @lafriks is this request how ready to go in? Or would you prefer that #5199 go in first and I update to match that?

**Comment by HoffmannP** [comment]
I only had to update the swagger file, now the CI built successfully

**Comment by zeripath** [comment]
Oh it's just a swagger update? Fine, no problem, I've been doing those pretty regularly anyway, (and then rebase on to master to keep the PR looking clean.) I was worried you had changed something internal to the API routing or that my routes needed seriously changing. 

I can't quite understand what your PR is doing over the web API, do you think there is anything else I need to do apart from the swagger update?

**Comment by sapk** [comment]
@zeripath the PR of @HoffmannP will already update the sdk deps. So you will maybe need to rebase to solve some conflict on deps and lock file but nothings to worry and no need to update the deps in this PR.

**Comment by zeripath** [comment]
@sapk I just had to remove my first commit as the dependencies had already been updated.

Is there anything else to be done for this request? I can remove the content searching if the query is thought to be too complex.

**Comment by lafriks** [comment]
To be honest I don't quite like that key searching by content. What is use case for that?

**Comment by zeripath** [comment]
Hi @lafriks ! The main use case for searching by key-content is creating a SSH AuthorizedKeysCommand [1][2] but you can do that using the fingerprint without the using the content. I'm happy to drop that bit or stick it behind an admin lock. (This is why the commits are separated rather than squashed in to one.)

I can create an AuthorizedKeysCommand command for the main gitea if you'd like.

[1] https://man.openbsd.org/sshd_config#AuthorizedKeysCommand
[2] https://man.openbsd.org/sshd_config#TOKENS

**Comment by techknowlogick** [comment]
I think it's fine to leave search by key in because keys are already available publicly (especially because they are "public keys"). I don't see this as a problem leaving in, and am supportive of the `AuthorizedKeysCommand` use case (I believe there is a ticket already for that).

File: integrations/api_keys_test.go
Comment (techknowlogick) [inline integrations/api_keys_test.go:124]: This tests fetching as a logged in user (due to token passing), can you also have a test for logged out users?

**Comment by zeripath** [comment]
@techknowlogick 
> I think it's fine to leave search by key in because keys are already available publicly (especially because they are "public keys"). I don't see this as a problem leaving in, and am supportive of the `AuthorizedKeysCommand` use case (I believe there is a ticket already for that).

@lafriks was referring to the search by key content+/-type code in https://github.com/zeripath/gitea/commit/e822116a0a07d6e5199e0aee95c2a537421853cf I don't know if you saw that. It's a little awkward due to the way Gitea stores key content at present, so in order to make it not possible to just list all the keys and their comments I put some more logic in.

The main reason for me putting in this PR is so that I can take a key-id and map it back to a user without having to iterate across all the users. I want to use that functionality to provide svn+ssh support for git-as-svn. That doesn't need the searching by key fingerprint or content, but putting that support in would allow me to write a separate AuthorizedKeysCommand rather than for me to have to parse and strip the authorized_keys files. Once you can create a AuthorizedKeysCommand you can provide other ssh based services on your own Gitea server.

For plain old gitea you can create an AuthorizedKeysCommand without this PR and I've just put a pull request in for an AuthorizedKeysCommand #5236 that doesn't need this (although if you pull this request we can add the fingerprint stuff in.)

File: integrations/api_keys_test.go
Comment (zeripath) [reply integrations/api_keys_test.go:124]: Do you mean testing the `/api/v1/user/:username/keys` routes?

**Comment by lafriks** [comment]
Does searching by content really requires searching by part of it or searching by equal value wouldn't be enough?

**Comment by zeripath** [comment]
@lafriks  Gitea is storing key type, content and comment as a single content string.

You don't get given the comment when you login from ssh so you can't just use equality. The function SearchPublicKeyByContent(content string) overloads the meaning of content in slightly different way by combining key type and content as a single string so it then uses a simplistic like, but if you just use that method you can retrieve all keys by passing in an empty content or 'ssh'. It also does no escaping of '%' or '_' in the like statement. 

(It's safe to use that function in the AuthorizedKeysCommand PR #5236 because SSHD will still match the keys individually.)

**Comment by zeripath** [comment]
@lafriks yeah that's fair enough. It's no longer in this PR.

File: integrations/api_keys_test.go
Comment (zeripath) [reply integrations/api_keys_test.go:124]: Just updated the integration test to test more paths.

**Review by lafriks — APPROVED** [review]

**Review by sapk — APPROVED** [review]

**Comment by techknowlogick** [comment]
@zeripath seems my email reply to the comment above didn't make it through, but it was just thanking your for clarifying.

**Pull request merged**
